add mysql row format dynamic option

peterhashair opened the pull request:
### Description:

Add row format equals dynamic as default, regarding the large text field for creating table failure. 
### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
remove html changes

Commit:
Update Settings.php

add key not exist

sgiehl commented on core/Db/Settings.php:
```diff
@@ -32,9 +32,14 @@ public function getUsedCharset()
 
+    public function getRowFormat()
+    {
+       return $this->getDbSetting('disable_dynamic_row_format')?'':'ROW_FORMAT=DYNAMIC';
```
The setting `disable_dynamic_row_format` doesn't exist in database configuration and would need to be added here: https://github.com/matomo-org/matomo/blob/7e0b0e602068a558a84f1e16fde5e7e6cb0b7fa4/config/global.ini.php#L43-L47

We can't simply use the dynamic row format for all instances, as some might break. So we need to disable that feature by default in `global.ini.php` and overwrite the value automatically if that feature is supported. This needs to be done while installing Matomo like here:
https://github.com/matomo-org/matomo/blob/09c012345d6913364af475598872105ff2fc854a/plugins/Installation/Controller.php#L608
and within a new updates script like this one https://github.com/matomo-org/matomo/blob/03410d62c6885010bc0968dca9ebe5e23b7c8e87/core/Updates/4.4.0-b1.php#L31-L36
To check if the feature is supported guess we could simply check if `DbHelper::getDefaultCharset()` returns `utf8mb4` or not.

Btw. Imho using some more spaces makes the code easier to read. e.g.
```php
return $this->getDbSetting('disable_dynamic_row_format') ? '' : 'ROW_FORMAT=DYNAMIC';
```

peterhashair replied:
thx sgiehL, updated it

Commit:
add disable_dynamic_row_format option into global config

update the database change on the Core Updates

tsteur commented on config/global.ini.php:
```diff
@@ -45,6 +45,9 @@
 charset = utf8
 
+; disable dynamic row by default
+disable_dynamic_row_format = true
```
fyi here we usually use `1` for enabled and `0` for disabled.

Also might be more readable by naming it `enable_dynamic_row_format = 0` but not too important

tsteur commented on core/Updates/4.5.0-b1.php:
```diff
@@ -29,11 +29,16 @@ public function __construct(MigrationFactory $factory)
+
+        //database using utf8mb4 set row format dynamic
+        if (DbHelper::getDefaultCharset() == "utf8mb4") {
+            $migrations[] = $this->migration->config->set('disable_dynamic_row_format', false);
```
fyi similar to above comment we usually would use here `0` instead of `false` for these config values.

tsteur commented on config/global.ini.php:
```diff
@@ -45,6 +45,9 @@
 ; Matomo should work correctly without this setting but we recommend to have a charset set.
 charset = utf8
 
+; disable dynamic row by default
```
could we maybe explain a bit more why someone would want dynamic rows? eg mention it avoids "row size too large errors" for new tables. or maybe we could also mention that "we recommend enabling this setting if the default charset is set to utf8mb4".

cc @sgiehl  does that make sense?

Commit:
update explain and content

convert naming from disable to enable, add comment, update update, take 3 params.

Commit:
fix the create table format ending space

There is an extra space on create sql, remove the ending space, if there is any. to Pass test.

Commit:
Update global.ini.php

update some description

tsteur commented on config/global.ini.php:
```diff
@@ -45,6 +45,10 @@
 ; Matomo should work correctly without this setting but we recommend to have a charset set.
 charset = utf8
 
+; Matomo recommend enabling this setting if the default charset is set to utf8mb4
```
can we slightly tweak this to `Matomo recommends enabling this setting if the default charset is set to "utf8mb4". It can avoid the MySQL error "row size too large" for new tables.`?

tsteur commented on core/Updates/4.5.0-b1.php:
```diff
@@ -29,11 +29,16 @@ public function __construct(MigrationFactory $factory)
+        $migrations[] = $this->migration->db->changeColumnType('session', 'data', 'MEDIUMTEXT');
+
+        //database using utf8mb4 set row format dynamic
+        if (DbHelper::getDefaultCharset() == "utf8mb4") {
```
@peterhashair because we have already release 4.5.0-b1 we would usually need to put this migration into a new file `4.5.0-b2`. It's not too important maybe for this one so we can leave it for now and don't need to change it.

peterhashair replied:
@tsteur Removed.

tsteur replied:
@peterhashair maybe you forgot to commit the new update fail? Not seeing the migration anymore anywhere

tsteur reviewed (commented):
@sgiehl I'm not quite sure what we do during the installation. Are we using utf8mb4 by default when possible? Would we need to do anything during the installation so users benefit from this?

Commit:
Merge branch '4.x-dev' into m-16834-row-format-mysql

Commit:
update global wording

update global wording and remove dynamic on in 4.5.0-b1

Commit:
add createTable dynamic test

add createTable dynamic test

tsteur commented:
@sgiehl be great to let us know if you know maybe whether we need to do anything on installation re this setting

sgiehl commented:
Yes. We need to use a default that works for all users and change the value on installation. Them same way it's done for utf8mb4

Commit:
add migration row format dynamic

add migration row format dynamic

Commit:
installation update

check when installation database is utf8mb4 format then auto enable  dynamic row format

peterhashair commented:
@tsteur Updated the Installation script, but not sure about the **ms sql** option, I set it `false` for now.

Commit:
update install mysql schema

update install mysql schema

Commit:
Merge branch '4.x-dev' into m-16834-row-format-mysql

Commit:
update unit test

update unit test

tsteur commented on core/Db/Settings.php:
```diff
@@ -32,9 +33,14 @@ public function getUsedCharset()
 
+    public function getRowFormat()
+    {
+        return $this->getDbSetting('enable_dynamic_row_format') ? 'ROW_FORMAT=DYNAMIC' : '';
```
@peterhashair I'm now actually wondering if we actually need this config setting? If I understand it now right, then we can apply that row format whenever `$this->getUsedCharset() === 'utf8mb4'`?  Is that correct @sgiehl  or would we still need a setting in case some DB doesn't support this row format?

peterhashair replied:
@tsteur  That's right, do you want me to remove the configuration, apply it to default if ut8mb4.

sgiehl replied:
In theory there might be cases where `ROW_FORMAT=DYNAMIC` might be supported, but `utf8mb4` collation is disabled manually. Not sure if anyone would do that or if that makes sense at all 🤷 
But guess for simplicity we could for now go without the config option and if there are coming too many complains that it breaks something we can introduce the config option later maybe?

tsteur commented on plugins/Installation/Controller.php:
```diff
@@ -149,6 +149,9 @@ function databaseSetup()
                 DbHelper::checkDatabaseVersion();
 
+                // check if database is UTF8MB4 
+                $dbInfos['enable_dynamic_row_format'] = Db::get()->isConnectionUTF8MB4() ? 1 : 0;
```
fyi I believe here we can simplify this using 
```php
        $config->database['enable_dynamic_row_format'] = $config->database['charset'] === 'utf8mb4' ? 1 : 0;
```

in the `createConfigFile` method. Then you won't need any of the methods using `isConnectionUTF8MB4` I suppose

peterhashair replied:
@tsteur that's right, let me tide this up.

Commit:
update dynamic to default when utf8mb4 is set

update dynamic to default when utf8mb4 is set

tsteur commented on core/Db/AdapterInterface.php:
```diff
@@ -70,4 +70,10 @@ public function isErrNo($e, $errno);
      */
     public function isConnectionUTF8();
+
+    /**
```
fyi this can be removed now

tsteur commented on core/Db/Schema/Mysql.php:
```diff
@@ -514,12 +514,14 @@ public function createTable($nameWithoutPrefix, $createDefinition)
     {
         $dbSettings   = new Db\Settings();
         $charset      = $dbSettings->getUsedCharset();
+        $dynamic = $charset==='utf8mb4' ? 'ROW_FORMAT=DYNAMIC' : '';
```
technically it could be nice to still have the row format logic in `$dbSettings->getRowFormat()` then it is more reusable but not too important.
the variable you maybe want to name `$rowformat` instead of `$dynamic`

tsteur commented on core/DbHelper.php:
```diff
@@ -160,6 +160,18 @@ public static function isDatabaseConnectionUTF8()
+     *
+     * @return bool  True if it is (or doesn't matter); false otherwise
+     */
+    public static function isDatabaseConnectionUTF8MB()
```
this one we'll no longer need 👍

tsteur commented on core/Updates/4.5.0-b2.php:
```diff
@@ -0,0 +1,44 @@
+use Piwik\Updates as PiwikUpdates;
+use Piwik\Updater\Migration\Factory as MigrationFactory;
+
+class Updates_4_5_0_b2 extends PiwikUpdates
```
this update script we can now remove as well now that we removed the config setting

tsteur reviewed (changes requested):
Nice, we're getting there. Looks a lot cleaner without the config 👍  sorry for having to remove it but it'll simplify things in the future 🚀

Commit:
remove uft8 function, set row format as a quick function, remove migration updates

remove uft8 function, set row format as a quick function, remove migration updates

peterhashair commented:
@tsteur Sorry about the issue being dragging too long, I should scope it a little bit better at the beginning.

tsteur commented:
@peterhashair all good, the problem was we didn't scope the issue well enough in the actual issue. We'll do better :) And good we found that we can simplify it all that's great.

Commit:
Merge branch '4.x-dev' into m-16834-row-format-mysql

# Conflicts:
#	core/Db/Adapter/Pdo/Mssql.php
#	core/Db/Adapter/Pdo/Pgsql.php

Commit:
resolve conflicts

resolve conflicts

sgiehl commented on core/Db/Schema/Mysql.php:
```diff
@@ -515,11 +515,12 @@ public function createTable($nameWithoutPrefix, $createDefinition)
         $charset      = $dbSettings->getUsedCharset();
 
-        $statement = sprintf("CREATE TABLE IF NOT EXISTS `%s` ( %s ) ENGINE=%s DEFAULT CHARSET=%s ;",
+        $statement = rtrim(sprintf("CREATE TABLE IF NOT EXISTS `%s` ( %s ) ENGINE=%s DEFAULT CHARSET=%s %s;",
```
guess the `rtrim` is actually obsolete, as the query always ends with a `;`

peterhashair replied:
@sgiehl Thanks for pointing that out. :)

sgiehl reviewed (commented):
left a minor comment. Otherwise looks good to me

Commit:
remove rtrim

remove rtrim obsolete

sgiehl reviewed (approved):
Looks good to me know
